Code/GraphMol/RWMol.cpp

```diff
@@ -574,26 +576,272 @@ void RWMol::beginBatchEdit() {
   dp_delAtoms.reset(new boost::dynamic_bitset<>(getNumAtoms()));
   dp_delBonds.reset(new boost::dynamic_bitset<>(getNumBonds()));
 }
+
 void RWMol::commitBatchEdit() {
-  if (!dp_delBonds || !dp_delAtoms) {
-    return;
-  }
-  auto delBonds = *dp_delBonds;
-  dp_delBonds.reset();
-  for (unsigned int i = delBonds.size(); i > 0; --i) {
-    if (delBonds[i - 1]) {
-      const auto bnd = getBondWithIdx(i - 1);
-      CHECK_INVARIANT(bnd, "bond not found");
-      removeBond(bnd->getBeginAtomIdx(), bnd->getEndAtomIdx());
+    if (!(dp_delBonds || dp_delAtoms)) {
+        return;
     }
-  }
-  auto delAtoms = *dp_delAtoms;
-  dp_delAtoms.reset();
-  for (unsigned int i = delAtoms.size(); i > 0; --i) {
-    if (delAtoms[i - 1]) {
-      removeAtom(i - 1);
+
+    batchRemoveBonds();
+    batchRemoveAtoms();
+    
+    // remove ring info
+    dp_ringInfo->reset();
+    
+    // fix properties
+    clearComputedProps(true);
+    dp_delBonds.reset();
+    dp_delAtoms.reset();
+}
+
+void RWMol::batchRemoveBonds() {
+    if(!dp_delBonds)
+        return;
+    
+    auto &delBonds = *dp_delBonds;
+    unsigned int min_idx = getNumBonds();
+    for (unsigned int i = rdcast<unsigned int>(delBonds.size()); i > 0; --i) {
+        if( !delBonds[i-1] )
+            continue;
+        unsigned int idx = rdcast<unsigned int>(i-1);
+        Bond *bnd = getBondWithIdx(idx);
+        if(!bnd) {
+            continue;
+        }
+        
+        min_idx = idx;
+        // remove any bookmarks which point to this bond:
+        BOND_BOOKMARK_MAP *marks = getBondBookmarks();
+        auto markI = marks->begin();
+        while (markI != marks->end()) {
+            BOND_PTR_LIST &bonds = markI->second;
+            // we need to copy the iterator then increment it, because the
+            // deletion we're going to do in clearBondBookmark will invalidate
+            // it.
+            auto tmpI = markI;
+            ++markI;
+            if (std::find(bonds.begin(), bonds.end(), bnd) != bonds.end()) {
+                clearBondBookmark(tmpI->first, bnd);
+            }
+        }
+        
+        // loop over neighboring double bonds and remove their stereo atom
+        //  information. This is definitely now invalid (was github issue 8)
+        ADJ_ITER a1, a2;
+        auto aid1 = bnd->getBeginAtomIdx();
+        auto aid2 = bnd->getEndAtomIdx();
+        boost::tie(a1, a2) = boost::adjacent_vertices(aid1, d_graph);
+        while (a1 != a2) {
```

Comment:
I think this could all be simplified (and possibly become faster) if you modify this to loop over the neighboring bonds with `atomBonds(aid1)`

You can probably also consolidate this loop and the duplicate one below it by having a single loop with this around it:
```
for(auto aid : {bnd->getBeginAtomIdx(), bnd->getEndAtomIdx()} ){
  for( auto obnd : atomBonds(aid) ) {
    if(obnd == bnd) {
       continue;
    }
```

Reply:
Nice.  I'll look into that.

Reply:
I'll update the original code as well if it works.

Reply:
I don't think it buys much here since we need to resolve both ends of the bond.  I tried the following which worked, let me know if you like this better and I'll merge it.

```
  // loop over neighboring double bonds and remove their stereo atom
  //  information. This is definitely now invalid (was github issue 8)
  auto beginAtm = bnd->getBeginAtom();
  auto endAtm = bnd->getEndAtom();
  std::vector<std::vector<Atom*>> bond_atoms ={{beginAtm, endAtm}, {endAtm, beginAtm}} ;
  for(auto atoms : bond_atoms){
    for( auto obnd : this->atomBonds(atoms[0]) ) {
        if(obnd == bnd) {
            continue;
        }
        if (std::find(obnd->getStereoAtoms().begin(), obnd->getStereoAtoms().end(),
                            atoms[1]->getIdx()) != obnd->getStereoAtoms().end()) {
            // github #6900 if we remove stereo atoms we need to remove
            //  the CIS and or TRANS since this requires stereo atoms
            if (obnd->getStereo() == Bond::BondStereo::STEREOCIS ||
                obnd->getStereo() == Bond::BondStereo::STEREOTRANS) {
                    obnd->setStereo(Bond::BondStereo::STEREONONE);
                }
            obnd->getStereoAtoms().clear();
            }
        }
  }

```

Reply:
That snippet looks fine to me. It doesn't look like `atoms` needs to be a vector of vectors though... you never use `atoms[1]`

Reply:
It's in the second std::find

Reply:
The reason there were two loops was that we were looking for any nbr bond from start that has end as a stereo atom and any nbr bond from end that has start as a stereo atom.

Reply:
Maybe we need ```for (auto nbrbond:  bondBonds(bond))``` :)